sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/uninstall/UninstallPlanFactory.java

```diff
@@ -155,4 +174,67 @@ Plan getPlan() {
     DeregisterStep getDeregisterStep() {
         return deregisterStep;
     }
+
+    /**
+     * Determines what resources need to be unreserved and builds {@link ResourceCleanupStep}s for each of them.
+     * The returned mapping groups the steps according to the hostname of the agent where they are located.
+     */
+    private static Map<String, Collection<ResourceCleanupStep>> getResourceCleanupStepsByAgent(
+            Optional<String> namespace, StateStore stateStore) {
+        Collection<Protos.TaskInfo> allTasks = stateStore.fetchTasks();
+        Set<String> taskIdsInErrorState = stateStore.fetchStatuses().stream()
+                .filter(taskStatus -> taskStatus.getState() == Protos.TaskState.TASK_ERROR)
+                .map(taskStatus -> taskStatus.getTaskId().getValue())
+                .collect(Collectors.toSet());
+
+        // Filter the tasks to those that have actually created resources. Tasks in an ERROR state which are also
+        // flagged as permanently failed are assumed to not have resources reserved on Mesos' end, despite our State
+        // Store still listing them with resources. This is because we log the planned reservation before it occurs.
+        Collection<Protos.TaskInfo> tasksNotFailedAndErrored = allTasks.stream()
+                .filter(taskInfo -> !(FailureUtils.isPermanentlyFailed(taskInfo)
+                        && taskIdsInErrorState.contains(taskInfo.getTaskId().getValue())))
+                .collect(Collectors.toList());
+
+        Map<String, Set<String>> dedupedResourceIdsByAgent = new HashMap<>();
+        for (Protos.TaskInfo taskInfo : tasksNotFailedAndErrored) {
+            String hostname;
+            try {
+                hostname = new TaskLabelReader(taskInfo).getHostname();
+            } catch (TaskException e) {
+                LOGGER.warn(String.format("Failed to determine hostname of task %s", taskInfo.getName()), e);
+                hostname = "UNKNOWN_AGENT";
+            }
+
+            Set<String> agentResourceIds = dedupedResourceIdsByAgent.get(hostname);
+            if (agentResourceIds == null) {
+                agentResourceIds = new HashSet<>();
+                dedupedResourceIdsByAgent.put(hostname, agentResourceIds);
+            }
+
+            agentResourceIds.addAll(ResourceUtils.getResourceIds(ResourceUtils.getAllResources(taskInfo)));
+        }
+
+        LOGGER.info("Configuring resource cleanup of {}/{} tasks across {} agents",
+                tasksNotFailedAndErrored.size(), allTasks.size(), dedupedResourceIdsByAgent.size());
+
+        // Map the resource IDs into ResourceCleanupSteps
+        return dedupedResourceIdsByAgent.entrySet().stream().collect(Collectors.toMap(
+                entry -> entry.getKey(),
+                entry -> entry.getValue().stream()
+                        .map(resourceId -> new ResourceCleanupStep(resourceId, namespace))
```

Comment:
I'm not a big fan fo inline constructors like this as they tend to make things pretty difficult to read. Won't block the PR on this though.